From 838da171f0e03776ff0f41ace858df32e1b5ba31 Mon Sep 17 00:00:00 2001
Message-Id: <838da171f0e03776ff0f41ace858df32e1b5ba31.1598619579.git.bristot@redhat.com>
In-Reply-To: <cover.1598619579.git.bristot@redhat.com>
References: <cover.1598619579.git.bristot@redhat.com>
From: Daniel Bristot de Oliveira <bristot@redhat.com>
Date: Wed, 22 Apr 2020 11:26:02 +0200
Subject: [RFC PATCH RT 08/17] sched/idle: Annotate the preempt disable from
 idle tasks

Idle runs with preempt disabled. It actually does not disable
preemption: the preemption is not enabled in the return from scheduler,
and so it is not disabled before calling back the scheduler.

Annotate the call and the return of the scheduler by switching the
reason why the preemption was disabled.

No functional changes, and the code is only added if preemption tracer
is enabled.

Signed-off-by: Daniel Bristot de Oliveira <bristot@redhat.com>
---
 kernel/sched/idle.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/kernel/sched/idle.c b/kernel/sched/idle.c
index b743bf38f08f..ba9300d1b8f8 100644
--- a/kernel/sched/idle.c
+++ b/kernel/sched/idle.c
@@ -290,7 +290,9 @@ static void do_idle(void)
 	smp_mb__after_atomic();
 
 	sched_ttwu_pending();
+	trace_preempt_switch_to_sched(CALLER_ADDR0, get_lock_parent_ip());
 	schedule_idle();
+	trace_preempt_switch_not_sched(CALLER_ADDR0, get_lock_parent_ip());
 
 	if (unlikely(klp_patch_pending(current)))
 		klp_update_patch_state(current);
-- 
2.26.2

